### DISCUSSION TITLE
Guidance on Multithreading

### DISCUSSION THREAD

**Discussion by b3nten** [jrouwe/JoltPhysics.js#218] [Q&A] [answered]
Hello, first of all, thank you for all your work on this engine - it's pretty impressive.

I'm looking at integrating it into a Three.js based game engine I'm working on but I'm a bit unclear on the multithreading semantics. I have no experience with Emscripten so following #134 & #110 was a bit tricky. Originally I was planning on running my physics inside of a worker and using a combination of `postMessage` and `SharedArrayBuffer` to communicate with the game thread. But that's a whole level of complexity I'd like to avoid if possible, especially as the API surface is large and async `postMessage` calls aren't great for writing gameplay code.

So I'm looking into using Jolt in multithreaded mode on supported devices, and I think I have a decent understanding now but there are some things I'm unclear on still. Please correct me if I'm wrong on anything.

1) Most calls to Jolt work the same with multithreading enabled. Creating/modifying/removing bodies etc. are the same with either mode. I'm a bit unclear if it's possible to offload the expensive work to the wasm / a thread without blocking on the main thread? Either using callbacks on completion of the work, or simply ignoring a return value. 
2) There are a handful of exceptions to 1
    - Contact listeners: callbacks run during the sim to modify bodies, and javascript callbacks cannot be passed to threads.
    - Buoyancy
    - Path constraints 
3) The current example with contact listeners is to provide a worker script that receives a pointer to a contactListener from the main thread and the IDs of certain bodies and then registers callbacks in the worker, which I'm assuming is loaded into each thread. Once the worker script is loaded and Jolt is initialised, there can be no more communication with the worker other than through shared memory.

If those are the only significant things that are affected by multithreading, only contact listeners seem critical for me to solve for with a general solution. In my engine, Behaviors can be added and removed to the world/actors at runtime which would correspond to adding or removing bodies, and they would need a way to register contact listeners - I can probably separate it into contact events that a) run during the physics step (for filtering/modifications), and b) events that run after the step (for gameplay logic). My first thought is a shared array of stringified functions for a, and another shared array for contacts that can be read after for b. But of course open to other ideas.

Are there any other things I need to consider with multithreading? Any significant blockers that could warrant a change in direction?

Cheers :)

**Comment by jrouwe** [comment]
> Most calls to Jolt work the same with multithreading enabled. Creating/modifying/removing bodies etc. are the same with either mode. 

Correct.

> I'm a bit unclear if it's possible to offload the expensive work to the wasm / a thread without blocking on the main thread? Either using callbacks on completion of the work, or simply ignoring a return value.

It is probably possible, but AFAIK it hasn't been attempted yet. I think you could create 1 worker thread to do all the communication with Jolt and communicate from the main thread with the worker thread via a SharedArrayBuffer. I would expect this to be quite cumbersome (you'd have to define some sort of communication format). I'm also not sure if it is needed since the most expensive thing is the simulation itself (aside maybe from things like mesh creation).

> Contact listeners: callbacks run during the sim to modify bodies, and javascript callbacks cannot be passed to threads.

Correct.

> Buoyancy

I think this could run on the main thread as well. It's basically a function you need to call every frame on every body that's in the water.

> Path constraints

If you use `PathConstraintPathJS` then yes as it is a 'callback', if you would use something like `PathConstraintPathHermite` (implemented in C++ and I see I forgot to expose the interface to JS) then everything stays in C++ and you would not need to worry about it.

> The current example with contact listeners is to provide a worker script that receives a pointer to a contactListener from the main thread and the IDs of certain bodies and then registers callbacks in the worker, which I'm assuming is loaded into each thread. Once the worker script is loaded and Jolt is initialised, there can be no more communication with the worker other than through shared memory.

I think that's correct too, although I'm not an expert on the JS side.

>  and they would need a way to register contact listeners

There's only 1 global contact listener. So you don't register new ones. Internally you would need to look at the bodies and decide what logic to execute for them.

> My first thought is a shared array of stringified functions for a, and another shared array for contacts that can be read after for b.

Does that mean you share JS code for the callback and let the other thread compile/execute it? I would probably define some sort of data format to communicate between threads. It would be less flexible, but probably a lot faster.

> Are there any other things I need to consider with multithreading? Any significant blockers that could warrant a change in direction?

Not that I'm aware of. But the multi threading code is pretty new so I don't know what will pop up.

**Reply by jrouwe** [reply]
I exposed [PathConstraintPathHermite](https://github.com/jrouwe/JoltPhysics.js/commit/f3f2eefb0c65b7cd501fda449869da227420775b).

**Reply by b3nten** [reply]
>I'm also not sure if it is needed since the most expensive thing is the simulation itself (aside maybe from things like mesh creation).

Would it be possible to import Jolt into a worker and use it for mesh creation, serialising it to a binary format and sharing it with the main thread?

> Does that mean you share JS code for the callback and let the other thread compile/execute it? I would probably define some sort of data format to communicate between threads. It would be less flexible, but probably a lot faster.

I haven't gotten to implementing contact listeners yet but I was under the impression that there are JS callbacks that run during the physics step to modify or accept/reject contacts. I'm unclear how communication between the main thread would work here - If this isn't the case then contact event data could just be shared with the main thread for calling callbacks after the physics step.

**Reply by jrouwe** [reply, marked as answer]
> Would it be possible to import Jolt into a worker and use it for mesh creation, serialising it to a binary format and sharing it with the main thread?

Yes, but `Shape::SaveBinaryState` has not been exposed yet.

> I haven't gotten to implementing contact listeners yet but I was under the impression that there are JS callbacks that run during the physics step to modify or accept/reject contacts. I'm unclear how communication between the main thread would work here - If this isn't the case then contact event data could just be shared with the main thread for calling callbacks after the physics step.

There's a single global JS callback that you register on PhysicsSystem. It can receive callbacks on all workers and in the callback it can use whatever state is available from C++ or it can use whatever state has been shared through a SharedArrayBuffer. In the example this shared state contains the body IDs of the things that the contact listener should be interested in. It can indeed also use that same SharedArrayBuffer to write stuff back to the main thread so that you can call callbacks at the end of the simulation step.

**Reply by b3nten** [reply]
> There's a single global JS callback that you register on PhysicsSystem. It can receive callbacks on all workers and in the callback it can use whatever state is available from C++ or it can use whatever state has been shared through a SharedArrayBuffer. In the example this shared state contains the body IDs of the things that the contact listener should be interested in. It can indeed also use that same SharedArrayBuffer to write stuff back to the main thread so that you can call callbacks at the end of the simulation step.

Ok cool thats what I thought. And I'm guessing there is no chance of cross-thread communication during those callbacks? In which case I think passing callbacks as strings to the worker via shared memory would be the only solution for runtime contact event registration (which is totally fine). I imagine most contact event handling would be more concerned with gameplay logic and would run after the step on the main thread.

Thank you for your help!

**Reply by jrouwe** [reply]
> And I'm guessing there is no chance of cross-thread communication during those callbacks?

Not in the sense that you can send a message to another thread and wait for a reply. But I think it's possible to lock a mutex and synchronize state in a SharedArrayBuffer across multiple workers in JS if I understand [this comment](https://github.com/jrouwe/JoltPhysics.js/issues/134#issuecomment-2425312299) correctly. I haven't done this myself and I would also avoid synchronization in those callbacks as that will slow down the simulation.

**Reply by PhoenixIllusion** [reply]
One of the biggest limitation is that when those contact listeners execute, the main thread will be in the Physics Step, and the main thread will not exit that function until the Step is done, so cannot receive worker messages in realtime. It would never have the chance to communicate with the workers.

The worker threads do not have any real chance to execute any code at all except during a callback, which happens only at the same time as the main threads Step(), and also have no guarantee that they will even run every Step, or ever receive a callback again in the lifetime of the app.

During a callback, You can atomic lock an index of a shared buffer array to reproduce mutex/semaphore/locking behavior, but I would recommend sticking to things like atomic-increment and other non-blocking functions. I would avoid ever trying to have the workers depend on each other or try to collaborate on anything in parallel given their execution is up to the job manager and any given thread actually triggering the callback.

They can however post back to the main thread that will be processed in the main threads next free "yield" at some later time, but workers will never receive a postMessage themselves as long as their Wasm thread is running.

I've thought of trying to make a collision tracker that could be put into my custom Jolt build's wasm written in C++, so I could possibly skip any Worker custom code in 95% of cases.

If you were to design a SharedArray based pattern to communicate with workers and gather collisions, you would need some schema, such that you could read the shared array and know what contacts to look for and how/where to save the data.

My fallback brainstorm for my plugin library was to store something in the array like:

[list of pairs of collisions to track, stored as a single u32 count + <count> userData id-pairs]
[atomic pointer to next write address in output buffer]
[the output buffer to write collision records]
